### ISSUE TITLE
False-positive with Checker Framework 3 and Buck on JDK11

### ISSUE THREAD

**Issue by swarren12** [typetools/checker-framework#5058]
Recently, I helped in an attmept to update a large monorepo built using Buck from JDK1.8 to JDK11; as part of that migration Checker Framework was updated from 2.10 to 3.17. However, the project no longer builds as Checker Framework throws errors that were not present previously. From what I can tell, it appears that the upgrade has introduced some false positives.

I have recreated one particular problem with a cutdown project.

At a guess I would point a finger here at some incompatibility between Checker 3 and Buck; however, [the manual](https://checkerframework.org/manual/#buck) does not seem to suggest that there are any known issues.

### Commands

```
export JAVA_HOME=/path/to/your/jdk11
git clone https://github.com/swarren12/checker-weirdness.git
cd checker-weirdness
./build-buck
./buck fetch //... && ./buck build //app:main
```

### Inputs

n/a

### Outputs

```
Command failed with exit code 1.

stderr: app/src/main/java/com/example/Main.java:11: error: [conflicting.annos] invalid type: conflicting annotations [@s(Prefix.milli), @s(Prefix.milli)] in type "@s(Prefix.milli) @s(Prefix.milli) long"
        final @Milliseconds long time = clock.currentTimeMillis();
                                                               ^
Errors: 1. Warnings: 0.

    When running <javac>.
    When building rule //app:main.
```

### Expectation

The project should build without an error.

***Justification***:
- the line that is failing is `final @Milliseconds long time = clock.currentTimeMillis();`.
- the `Clock` interface is annotated as `@Milliseconds long currentTimeMillis();`;
- therefore the signatures should match.

Additionally, if I'm reading the error correctly, Checker Framework _does think the signatures match_ and so I'm unclear as to exactly why it is failing.

**Comment by mernst** [comment]
Thanks for the bug report.  I'm sorry you are having trouble.

I appreciate the simple reproduction instructions.  I am able to reproduce your problem within Buck.

However, I have not been able to reproduce the problem outside Buck.
The two attached variants of your project show three different attempts.  (If you are interested, see the two README files for details.)

How can I get Buck to output the exact javac command line that it runs, so that I can reproduce the problem outside Buck?
I don't see documentation of this at, for example, https://buck.build/command/build.html (except for C++, and those instructions don't work in your build).

[checker-weirdness-2.tar.gz](https://github.com/typetools/checker-framework/files/8111780/checker-weirdness-2.tar.gz)
[checker-weirdness-3.tar.gz](https://github.com/typetools/checker-framework/files/8111778/checker-weirdness-3.tar.gz)

**Comment by swarren12** [comment]
I believe running the build with `--verbose 8` should give you the output you want, i.e. `./buck build --verbose 8 //app:main`. You may need to do a `./buck clean` first to get some meaningful output.

**Comment by swarren12** [comment]
I ran this myself earlier and I've attached the [output](https://github.com/typetools/checker-framework/files/8115831/err.log). Specifically, the command that appears to be failing is:
```
javac -source 11 -target 11 -sourcepath  -g -processor org.checkerframework.checker.units.UnitsChecker -processorpath buck-out/gen/__checker-framework__/checker-framework.jar:buck-out/gen/__checker-qual__/checker-qual.jar:buck-out/gen/__checker-util__/checker-util.jar -verbose -d buck-out/bin/app/lib__main__scratch/classes -s buck-out/annotation/app/__main_gen__ -classpath buck-out/gen/__checker-framework__/checker-framework.jar:buck-out/gen/__checker-qual__/checker-qual.jar:buck-out/gen/__checker-util__/checker-util.jar:buck-out/gen/annotations/time/lib__annotations-time__output/annotations-time.jar:buck-out/gen/utils/clock/lib__clock__output/clock.jar @buck-out/gen/app/__main__srcs
```

Some other observations: 
- Running this command from the terminal doesn't reproduce the error; so either Buck isn't being entirely honest about what it's running or there's some difference in the `javac` being used by Buck and being used outside?
- I added a new branch to the project showing that Bazel doesn't suffer the same problem, further pointing to Buck doing something weird!

**Comment by msridhar** [comment]
This is weird.  My wild guess would be some classloader issue, where the `@Milliseconds` annotation is somehow getting loaded by two unrelated classloaders and hence they look incompatible to the Checker Framework.  But if the build succeeds after multiple runs then it might be something different.

@artempyanykh any idea what is going on here (if you have time)?

**Comment by artempyanykh** [comment]
Assuming, there's not a lot of differences between our internal setup and OSS version of buck (I think there shouldn't, at least conceptually):
1. The command lines that buck prints with `-v 8` is not what's actually getting executed.
2. Buck doesn't invoke external javac, but uses in process compilation [[link](https://github.com/facebook/buck/blob/main/src/com/facebook/buck/jvm/java/Jsr199Javac.java)]. This is configurable to an extent [[docs](https://buck.build/files-and-dirs/buckconfig.html#tools.javac)].
3. It also may use multiple classloaders that are cached [[link](https://github.com/facebook/buck/blob/main/src/com/facebook/buck/util/ClassLoaderCache.java)].

When I need to debug such issues in our code, I usually run buck in debug mode (it's java process anyway), put breakpoints in the analyser's code and see what's going on exactly that makes the check fail.

**Comment by mernst** [comment]
@artempyanykh Is there a way to configure Buck to use an external javac?  In general I realize that starting a new process is inefficient.

**Comment by mernst** [comment]
I have made a pull request that summarizes this discussion in the Checker Framework manual.

Could one of you please review it to ensure that it is accurate?
Please see https://github.com/typetools/checker-framework/pull/5072 .

Thanks!
